### ISSUE TITLE
How to do a stable automated cluster reboot

### ISSUE THREAD

**Issue by dyoakley** [ClusterLabs/PAF#147]
3 node cluster, 1 master and 2 slaves (one synchronous standby, one async standby), I'm looking for a way reboot all my nodes and have them come back online automatically and quickly, ideally with master in the same place.  My first thought was: 
1) put the cluster in maintenance mode so there is moving around on shutdown
2) reboot all nodes in parallel
3) bring up pacemaker
4) wait long enough time for all nodes to be back online and to join corosync network and have a quorum
5) take out of maintenance mode (to let the agent bring up the instances again)

Sometimes that goes okay, master does not move and comes up immediately.  Other times the master gets immediately put in standby on step 5, a 15 minute delay and then master gets promoted again....  

I realize I'm swimming upstream using PAF to build an automated appliance that allows pacemaker to come up on its own etc.  It's been a bit rocky but actually worked out okay.  Hoping someone might give me some insight as to a way to do this reliably...

**Comment by ioguix** [comment]
Hello,

The proper way to do it is:

1. disable your master/promotable resource, eg.: `pcs resource disable pgsql-ha`
2. shutdown the cluster, eg.: `pcs cluster stop --all`
3. reboot all your nodes
4. start Pacemaker, eg.: `pcs cluster start --all`
5. start the master/promotable resource, eg.: `pcs resource enable pgsql-ha`

(based on https://clusterlabs.github.io/PAF/Quick_Start-CentOS-7.html)

When disabling the pgsql-ha master resource, it keeps the master score around and all clones are stopped in one transition. No matter what happen during the Pacemaker shutdown.

To answer your question, when Pacemaker starts a multi-state resource, it MUST always start it as a standby first, then it promotes it. If something goes messy and some condition breaks the active transition, it is sometimes not able to finish all the jobs up to the promotion. 15 minutes later, the `cluster-recheck-interval`triggers a global check of the cluster and find a node could be promoted.

Anyway, try with my procedure and report if you still have troubles.

**Comment by dyoakley** [comment]
Great thanks!  BTW,  It looks like when I hit the delay I was coming out of maintenance mode before the DC is established.  If I wait until after that, it seems like master is established right away.  I guess that means PAF was able to decide on the first check so I didn't get into the cluster-recheck-interval??

**Comment by dyoakley** [comment]
Thanks for the quick reply.  Is there a `crm` equivalent of `resource disable`?  Perhaps `resource unmanage`?

**Comment by dyoakley** [comment]
I found `crm cluster wait_for_startup` and was hopeful it would do the trick for me.  It released too early as well, allowing me to come out of maintenance mode but still throwing off the agent so that it backed off for 15 minutes before I got a master.  That's a shame as this really seemed like the right tool to use to trigger on in terms of automation.  On the other hand, just hard coding a sleep long enough for the DC to update got me working again.  I don't like it much as obviously this should be event based.  As to "what" to trigger (maint mode false vs- resource enable) I'll wait to see what you think about the right way to disable/enable a resource with crm.  It seems to me with that approach I'm likely to run into this same issue, triggering agent into action before things are ready.

**Comment by ioguix** [comment]
>  Is there a crm equivalent of resource disable? Perhaps resource unmanage?

No, unmanage has a different meaning. Don't do that. The equivalent is `crm resource stop`.

I'm not sure of the capabilities of `crmsh`, I switched to pcs long time ago and sticked to it. You migh want to check if pcs is available on your distro. I know it works good under Debian. Using pcs, you can do `pcs cluster start --all --wait`.

But if you can't rely on neither pcs or crmsh, you can use `crmadmin`. eg. Using an infinite loop, a full cluster startup looks like this:

~~~
[root@srv1 ~]# while true; do sleep 0.5; echo -e "$(date): $(crmadmin -D)\n $(crmadmin -S srv1)\n $(crmadmin -S srv2)\n $(crmadmin -S srv3)"; done
...
Sat Jan 26 09:49:10 UTC 2019:
Sat Jan 26 09:49:11 UTC 2019:
No messages received in 30 seconds.. aborting
Sat Jan 26 09:49:11 UTC 2019: 
 Status of crmd@srv1: S_NOT_DC (ok)
 Status of crmd@srv2: S_IDLE (ok)
 Status of crmd@srv3: S_NOT_DC (ok)
Sat Jan 26 09:49:43 UTC 2019: Designated Controller is: srv2
 Status of crmd@srv1: S_NOT_DC (ok)
 Status of crmd@srv2: S_IDLE (ok)
 Status of crmd@srv3: S_NOT_DC (ok)
Sat Jan 26 09:49:43 UTC 2019: Designated Controller is: srv2
 Status of crmd@srv1: S_NOT_DC (ok)
 Status of crmd@srv2: S_IDLE (ok)
 Status of crmd@srv3: S_NOT_DC (ok)
...
~~~

When starting up the pgsql-ha master resource:

~~~
Sat Jan 26 09:56:52 UTC 2019: Designated Controller is: srv2
 Status of crmd@srv1: S_NOT_DC (ok)
 Status of crmd@srv2: S_IDLE (ok)
 Status of crmd@srv3: S_NOT_DC (ok)
Sat Jan 26 09:56:52 UTC 2019: Designated Controller is: srv2
 Status of crmd@srv1: S_NOT_DC (ok)
 Status of crmd@srv2: S_TRANSITION_ENGINE (ok)
 Status of crmd@srv3: S_NOT_DC (ok)
...
Sat Jan 26 09:56:58 UTC 2019: Designated Controller is: srv2
 Status of crmd@srv1: S_NOT_DC (ok)
 Status of crmd@srv2: S_TRANSITION_ENGINE (ok)
 Status of crmd@srv3: S_NOT_DC (ok)
Sat Jan 26 09:56:59 UTC 2019: Designated Controller is: srv2
 Status of crmd@srv1: S_NOT_DC (ok)
 Status of crmd@srv2: S_IDLE (ok)
 Status of crmd@srv3: S_NOT_DC (ok)
~~~

So creating a waiting loop around `crmadmin` and parsing it might do the trick for you here. Just wait for the DC to be up and S_IDLE and other nodes to be up and S_NOT_DC.

If you detect the cluster is idle AND your PostgreSQL resource is up but as a standby, instead of waiting for the recheck, you can force the promotion with the following command: 

`crm_master --node=srv1 --resource=pgsqld --update=2000`

Note that you have to give the PostgreSQL resource (here `pgsqld`), not the master resource associated with it (`pgsql-ha` in my examples). And make sure to pick the good node!

**Comment by ioguix** [comment]
Hi,

The consequences for lowering the cluster-recheck-interval too much is a lot of useless network communications, flood in log files and CPU consumption.

As instance, 1min is a nogo to me. 3 minutes sounds ok to me.
